proposals/streams/microprofile-streams-example/src/main/java/org/eclipse/microprofile/streams/example/UserDetailsSubscriber.java

```diff
@@ -0,0 +1,45 @@
+package org.eclipse.microprofile.streams.example;
+
+import akka.stream.Materializer;
+import akka.stream.javadsl.JavaFlowSupport;
+import akka.stream.javadsl.Sink;
+import org.eclipse.microprofile.streams.Envelope;
+import org.eclipse.microprofile.streams.Ingest;
+import org.eclipse.microprofile.streams.example.models.UserDetails;
+import org.eclipse.microprofile.streams.example.models.UserEvent;
+import org.eclipse.microprofile.streams.example.repository.UserDetailsRepository;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Initialized;
+import javax.enterprise.event.Observes;
+import javax.inject.Inject;
+import java.util.concurrent.Flow;
+
+@ApplicationScoped
+public class UserDetailsSubscriber {
+
+  private final UserDetailsRepository repository;
+  private final Materializer materializer;
+
+  @Inject
+  public UserDetailsSubscriber(UserDetailsRepository repository, Materializer materializer) {
+    this.repository = repository;
+    this.materializer = materializer;
+  }
+
+  @Ingest(topic = "user-details.events")
+  public Flow.Subscriber<Envelope<UserEvent>> handleUserDetails() {
```

Comment:
This method is in fact a producer of a topic subscriber, correct? Is it called once or multiple times for a topic?
If only once, then it's technically a producer of an application scoped bean of type Flow.Subscriber and the behavior should be the same as if it was a proper producer like this:

```
@javax.enterprise.inject.Produces
@Ingest(topic = "user-details.events")
public Flow.Subscriber<Envelope<UserEvent>> handleUserDetails() {
```
I like the idea of supporting the version without `@Produces` annotation so that it doesn't strictly require CDI, but in a CDI container the behavior of both versions should be equal - it simplifies the specification.

Reply:
Alternatively, I suggest also supporting a simpler API that doesn't require Flow API, like this:

```
@Ingest(topic = "user-details.events")
public CompletionStage<Void> handleUserDetails(Envelope<UserEvent> msg) {
    return repository.updateUserDetails(new UserDetails(event.getId(),
        event.getName(), event.getEmail()));
```

or with a void return value:

```
@Ingest(topic = "user-details.events")
public void handleUserDetails(Envelope<UserEvent> msg) {
   // do something and forget
```

It would be for very simple usecases when we don't need to care about errors or end of streams, only about next messages.

Reply:
Optionally, the above methods could receive Flow.Subscription as an additional argument to control the flow. Without the argument, `subscription.request(1)` would be called implicitly after the method is called. It would be also called automatically before first message received.

```
@Ingest(topic = "user-details.events")
public void handleUserDetails(Envelope<UserEvent> msg, Flow.Subscription subscription) {
   // do something and forget
   if (allHandled) {
      subscription.request(10);
   }
}
```

Reply:
@OndrejM I thought @jroper mentioned the scope is bound to the bean's scope. The reason to return a CompletionStage is for asynchronous, which makes the message deliver asynchronously.

Reply:
The method may be called multiple times - if there is an error in the stream processing, it will be invoked again to get a new `Subscriber`. `Subscriber`'s are by design single shot, so can't be reused.

I think we should investigate supporting the `Envelope<Message> -> CompletionStage<Void>` method, or even the `void` version of the API, for many use cases that will be simpler, and I think especially in the state that reactive streams is right now (where you really do need to bring in a third party streams implementation as I have with Akka Streams), such an API may be essential to successful adoption of this. I've mentioned briefly before, but we're currently working with some people on the JDK side to introduce a new API for building reactive streams graphs with an API that looks very similar to java.util.stream, but even if all the stars align and that lands in JDK11 (which it won't), that only further pushes up the baseline for such an API.

We can of course support both approaches.